### PR TITLE
Allow sample track TCOs to resize smaller than one bar

### PR DESCRIPTION
**Pull request by Spekular**
Currently, TCOs in sample tracks can't be resized to less than one bar, which is frustrating if short samples are loaded (they'll have useless empty space to the right). This PR allows sample tracks TCOs to be resized to less than one bar's size. Empty TCOs will still be created with a size of one bar.

At 100% zoom, it becomes difficult to see the TCO if it is at minimum size, but this already occurs with BB Track TCOs, so I wouldn't consider it a regression.

### CODE REVIEW

**Comment by Spekular** [comment]
Looks like some weird unrelated stuff ended up here, will fix.

File: src/tracks/SampleTrack.cpp [hunk 1]
```diff
@@ -147,8 +144,19 @@ void SampleTCO::setSampleBuffer( SampleBuffer* sb )
...
-	m_sampleBuffer->setAudioFile( _sf );
-	changeLength( (int) ( m_sampleBuffer->frames() / Engine::framesPerTick() ) );
+	int length;
+	if (_sf == "") //When creating an empty sample pattern make it a bar long
+	{
+		float nom = Engine::getSong()->getTimeSigModel().getNumerator();
+		float den = Engine::getSong()->getTimeSigModel().getDenominator();
+		length = DefaultTicksPerTact * ( nom / den );
+	}
+	else //Otherwise set it to the samples length
```
Comment (Spekular) [inline src/tracks/SampleTrack.cpp]:
```suggestion
	else //Otherwise set it to the sample's length
```

**Comment by BaraMGB** [comment]
Do you really want to change this for stable?

**Comment by Spekular** [comment]
I think you could argue the original behavior is a bug. I did intentionally target it for stable but if there are objections I could retarget to master.

**Comment by JohannesLorenz** [comment]
The freshly loaded samples (doubleclick on the sample track TCOs) load with a correct length, but when you resize them to 1, you can't resize them back. In this situation, even when you load the same sample file again, the TCO size is still 1. Both can cause confusion and may be worth a fix.

As for the stable-1.2 question: I'm fine with merging it into stable-1.2 after the review will have approved this branch.

**Comment by Spekular** [comment]
Hmm, I definitely tried scaling them up from zero and it seemed to work fine for me. How exactly do I reproduce the issue? Does the drag handle show up? Does it work when zoomed in further?

**Comment by JohannesLorenz** [comment]
1. Click on the sample track to create a yellow, empty TCO
2. Double click on that TCO, open a file
3. Move mouse on border of TCO, resize dialog appears
4. Resize to 1
5. Try to resize to <1 at any zoom, drag handle will shop up, but you can't make it smaller than 1

**Comment by Spekular** [comment]
@JohannesLorenz can't reproduce with those steps. Are you holding ctrl to fine adjust? Since it resizes in bar increments by default, 1 bar is the smallest it can go without fine adjusting.

The clip scales to the correct size when a sample is loaded for me too. However, if you attempt to reload the same sample, this doesn't happen. I think the call to setSampleFile might be skipped when loading the existing sample again? I'll look into that.

**Comment by Spekular** [comment]
Reformatted for brevity, but here's what I think the cause is. Will fix.
```
if( af != "" && af != m_tco->m_sampleBuffer->audioFile() )
	{ m_tco->setSampleFile( af ); }
```

**Comment by Spekular** [comment]
The sample track clip should now resize to the correct length no matter what sample is loaded (a new one or the same that it already contained).

**Comment by Spekular** [comment]
There's currently a bug with cloning. A short (less than one bar) sample clip cannot be cloned to the right of itself unless another sample clip exists between it and the clone's destination. In other words, a short sample at position A can be cloned to position C if B exists, but B cannot be cloned to position C.

A B [c]

**Comment by Spekular** [comment]
A TrackContentWidget::dragEnterEvent is fired when a sample clip is Ctrl+dragged over or off of another sample clip. For a clip of 1 bar or more, we exit the clip's "airspace" directly into another bar than the original clip. For a short clip, we end up in the same bar.

tcoPos is apparently only accurate to one bar, so it looks like we're pasting to the same position when pasting a short clip to the right of itself. If the clip is dragged to another track and back to the original, or on top of anotoer clip in the same track and back, tcoPos is forced to update again, which can allow short clips to be cloned to their right.

**Comment by Spekular** [comment]
Essentially short clips often evaluate true on ``tcoPos == grabbedTCOTact`` because tcoPos is only accurate to one bar and is only updated on some events. This returns false in canPasteSelection.

**Comment by Spekular** [comment]
Previous issue is fixed.

Allowing short sample clips exposes an edge case. Cloning a sample clip to a bar containing a short clip can completely cover the short clip. Whether or not a safeguard should be implemented for this is up for debate.

**Comment by JohannesLorenz** [comment]
@Spekular I did a few tests. While everything worked fine it's a bit sad that you can't snap two short blocks together. Though I don't see that much use, as the sample TCO lengths/positions can't be set to an exact value anyways. I'd say this may be a nice feature, but it's absolutely not stopping this PR. If you want, you can try to add this to this PR, but I see it as totally optional.

**Comment by Spekular** [comment]
@johanneslorenz I'd rather leave it as is and then take a crack at #874 once this is merged.

**Comment by Spekular** [comment]
@JohannesLorenz fixed everything except the tick_t comment. I wouldn't know how to implement that one, the scope seems a bit big.

File: src/tracks/SampleTrack.cpp [hunk 2]
```diff
@@ -147,8 +144,19 @@ void SampleTCO::setSampleBuffer( SampleBuffer* sb )
...
+	if (_sf == "")
+	{	//When creating an empty sample pattern make it a bar long
+		float nom = Engine::getSong()->getTimeSigModel().getNumerator();
+		float den = Engine::getSong()->getTimeSigModel().getDenominator();
+		length = DefaultTicksPerTact * ( nom / den );
+	}
+	else
+	{	//Otherwise set it to the sample's length
+		m_sampleBuffer->setAudioFile( _sf );
+		length = sampleLength();
```
[the remark this answers is not in the document]
Comment (Spekular) [reply src/tracks/SampleTrack.cpp]: :)

**Comment by JohannesLorenz** [comment]
I'm OK with not fixing the `tick_t`. Can you please [re-request review](https://github.blog/changelog/2019-02-21-re-request-review-on-a-pull-request/), so I can mark it as OK? I'm curious if that works :smile:

Note, before merging, maybe it's good to squash those 10 commits.

File: src/tracks/SampleTrack.cpp [hunk 3]
```diff
@@ -147,8 +144,19 @@ void SampleTCO::setSampleBuffer( SampleBuffer* sb )
 
 void SampleTCO::setSampleFile( const QString & _sf )
 {
-	m_sampleBuffer->setAudioFile( _sf );
-	changeLength( (int) ( m_sampleBuffer->frames() / Engine::framesPerTick() ) );
+	int length;
```
[the remark this answers is not in the document]
Comment (Spekular) [reply src/tracks/SampleTrack.cpp]: Skipping this as scope seems big.

**Comment by Spekular** [comment]
I on't have that option, perhaps because 

> Pull request authors can't request reviews unless they are either a repository owner or collaborator with write access to the repository.

I've done my best to squash as well, no idea if it worked.

**Comment by JohannesLorenz** [comment]
OK, It's accepted now, anyways. I think you did not squash it correctly: You squashed (a091d50) , but then merged (062041f) your unsquashed branch on top? I think you can solve it all by one simple `git rebase -i 97d5529c18bee51d638607a192ab91b131d41086`, marking all `s`, and no further merge.

**Comment by Spekular** [comment]
When I try that, I immediately get a merge conflict :)

**Comment by Spekular** [comment]
I ran into a brand new error on this squash attempt, aside from merge conflicts some commits couldn't be applied because ???, and I couldn't really tell how to proceed. Maybe whoever merges can do a squash and merge?

**Comment by PhysSong** [comment]
@JohannesLorenz Using the `Squash and merge` feature will be perfectly fine in this case. You may want to edit the commit message to remove redundant messages though.

File: src/tracks/SampleTrack.cpp [hunk 4]
```diff
@@ -440,8 +448,15 @@ void SampleTCOView::mouseReleaseEvent(QMouseEvent *_me)
 void SampleTCOView::mouseDoubleClickEvent( QMouseEvent * )
 {
 	QString af = m_tco->m_sampleBuffer->openAudioFile();
-	if( af != "" && af != m_tco->m_sampleBuffer->audioFile() )
-	{
+
+	if ( af.isEmpty() ) {} //Don't do anything if no file is loaded
+	else if ( af == m_tco->m_sampleBuffer->audioFile() )
+	{	//Instead of reloading the existing file, just reset the size
```
Comment (DomClark) [inline src/tracks/SampleTrack.cpp]: If a user modifies one of their audio files (not entirely unlikely; could be new vocals, updated bounced track, new export from another DAW, etc.), opening the file dialog and re-selecting the sample will no longer load the modified file. IMO this is unintuitive.

Comment (Spekular) [reply src/tracks/SampleTrack.cpp]: I didn't introduce this behavior. Existing version skips reloading entirely, my patch makes it skip reloading but at least reset size

Should i remove the check entirely?

Comment (DomClark) [reply src/tracks/SampleTrack.cpp]: Ah, indeed. I didn't read the original carefully enough, just saw the comment added and assumed it was new. In that case, I think the condition ought to go, but I don't know whether it's in scope here. Sorry.

Comment (Spekular) [reply src/tracks/SampleTrack.cpp]: I'd rather leave as is then since this targets stable.

**Pull request merged**